File: src/main/scala/rros/RROSProtocol.scala
package rros

import rros.core.RROSProtocolImpl

////////////////////////////////////////////////////////////////////////////////
/**
 * Request/Response over Socket Session 
 *
 * Created by namnguyen on 3/1/15.
 */
trait RROSProtocol extends AutoCloseable{
  //----------------------------------------------------------------------------
  def send(request: Request
           ,onComplete:(Response)=>Unit
           ,timeOut:Long = 30000 //30 seconds
           ,onFailure:(Exception) => Unit = (exc:Exception) => {
              //default log here
            }
  ):Unit
  //----------------------------------------------------------------------------
  def send(message:Message):Unit
  //----------------------------------------------------------------------------
  def send(rawString:String):Unit
  //----------------------------------------------------------------------------
  def onMessageReceived(callback:Option[(Message)=>Unit]):Unit
  def onRequestReceived(callback:Option[(Request)=>Response]):Unit
  def onAnythingReceived(callback:Option[(String)=>Unit]):Unit
  //----------------------------------------------------------------------------
}
////////////////////////////////////////////////////////////////////////////////
object RROSProtocol {
  def apply(socketAdapter: SocketAdapter) = new RROSProtocolImpl(socketAdapter)
}
////////////////////////////////////////////////////////////////////////////////

File: src/main/java/rros/java/RROSProtocol.java
package rros.java;

/**
 * Created by namnguyen on 3/9/15.
 */
public interface RROSProtocol {
    void send(Request request,SentAction action,long timeout);
    void send(Message message);
    void onMessageReceived(MessageReceivedAction action);
    void onRequestReceived(RequestReceivedAction action);

    public static interface SentAction {
        void onComplete(Response response);
        void onFailure(Exception exc);
    }
    public static interface MessageReceivedAction {
        void onReceived(Message message);
    }
    public static interface RequestReceivedAction {
        Response onReceived(Request request);
    }
}
